src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -26,10 +26,13 @@ namespace Microsoft.CodeAnalysis;
 [DebuggerDisplay("{GetDebuggerDisplay(),nq}")]
 public partial class Project
 {
-    private ImmutableDictionary<DocumentId, Document?> _idToDocumentMap = ImmutableDictionary<DocumentId, Document?>.Empty;
-    private ImmutableDictionary<DocumentId, SourceGeneratedDocument> _idToSourceGeneratedDocumentMap = ImmutableDictionary<DocumentId, SourceGeneratedDocument>.Empty;
-    private ImmutableDictionary<DocumentId, AdditionalDocument?> _idToAdditionalDocumentMap = ImmutableDictionary<DocumentId, AdditionalDocument?>.Empty;
-    private ImmutableDictionary<DocumentId, AnalyzerConfigDocument?> _idToAnalyzerConfigDocumentMap = ImmutableDictionary<DocumentId, AnalyzerConfigDocument?>.Empty;
+    private readonly object _gate = new();
+
+    // Only access these dictionaries when holding _gate
```

Comment:
alternatively, they could be their own locks.  but i don't care that much.  this would also enable concurrent access to all of them, while each one was threadsafe.  up to you.

Reply:
oh.  nevermind.  you use null. ignore this.

Reply:
Went ahead and switched the access of the dictionaries to be protected by using themselves as locks. Don't love it, but it does give better lock granularity.